dbms/src/Common/getNumberOfLogicalCPUCores.h

```diff
@@ -14,6 +14,18 @@
 
 #pragma once
 
-/// Get number of CPU cores without hyper-threading.
-/// Note: do not support environment under resource isolation mechanism like Docker, CGroup.
-unsigned getNumberOfPhysicalCPUCores();
+#include <common/types.h>
+
+inline UInt16 getNumberOfLogicalCPUCores(UInt16 max_logical_cpu_cores = 0)
+{
+    static UInt64 n = max_logical_cpu_cores;
+    return n;
+}
```

Comment:
I think it is dangerous to set the static var in function `getNumberOfLogicalCPUCores`. For example, creating a `SettingMaxThreads` will call `getNumberOfLogicalCPUCores(0)` by default, which unexpectedly makes it to be 0.

https://github.com/pingcap/tiflash/blob/269e77b1b5697af03e3f4f40408bf2ba3fab336f/dbms/src/Interpreters/SettingsCommon.h#L109-L112

Can we get the correct threads number from global_settings now?

Reply:
we declare global_context after setNumberOfLogicalCPUCores ?

Reply:
> I think it is dangerous to set the static var in function `getNumberOfLogicalCPUCores`. For example, creating a `SettingMaxThreads` will call `getNumberOfLogicalCPUCores(0)` by default, which unexpectedly makes it to be 0.
> 
> https://github.com/pingcap/tiflash/blob/269e77b1b5697af03e3f4f40408bf2ba3fab336f/dbms/src/Interpreters/SettingsCommon.h#L109-L112
> 
> Can we get the correct threads number from global_settings now?

maybe we can throw exception when the returned value is 0?